### PR TITLE
ui: expose proxy configuration in advanced configuration

### LINKED ISSUE #2093: UI: expose `spec.proxy.config` in the Advanced Configuration step
> Parent: #2092 
> Repo: this one (`openeverest/openeverest`)

## Context

Today the wizard's **Advanced Configuration** step lets users provide an
"Engine Configuration" snippet (e.g. `postgresql.conf`-style settings,
`mongod.conf` YAML, etc.) which Everest forwards to the upstream operator. We
need the same affordance for the **proxy** (mongos / pgBouncer / HAProxy /
ProxySQL).

The operator side is being implemented in the sibling sub-issues
(#2094, #2095; for PXC it already works.

### 1. Wizard form

- Add a new text area for proxy configuration directly underneath the
  existing "Engine Configuration" text area in
  `src/components/cluster-form/advanced-configuration/advanced-configuration.tsx`.
  The label must be **engine-specific**:
  - For PSMDB: **"Router Configuration"**
  - For PG: **"PG Bouncer Configuration"**
  - For PXC: **"Proxy Configuration"**
- Use the **exact same UX** as engine config: an enable/disable toggle
  (`proxyConfigEnabled`) followed by a multiline `<TextInput>` (`proxyConfig`)
  with placeholder + helper text. Reuse the label mapping from `getProxyUnitNamesFromDbType`
  in `src/utils/db.tsx` (append " Configuration" to the singular form).
- Add the two new fields to:
  - `advanced-configuration.types.ts` (`AdvancedConfigurationFields` enum)
  - `advanced-configuration-schema.ts` (Zod: `z.boolean()` and
    `z.string().optional()`)
  - `messages.ts` next to the component (label / description / placeholder)

### 2. Sharding gate (PSMDB only)

PSMDB only has a proxy when sharding is enabled. The new field must be hidden
when:
```
dbType === DbType.Mongo && !shardingEnabled
```
Reuse the existing pattern from
`steps/resources/resources-step.tsx`
(`hideProxies={dbType === DbType.Mongo && !shardingEnabled}`).

Pipe a new prop `showProxyConfig` from the wizard wrapper
`steps/advanced-configurations/advanced-configurations.tsx`
into `AdvancedConfigurationForm`.

### 3. Read/write mapping

- In `src/components/cluster-form/advanced-configuration/advanced-configuration.utils.ts`,
  read `dbCluster.spec.proxy.config` into the form's `proxyConfig` value (and
  set `proxyConfigEnabled` accordingly).
- In `src/utils/db.tsx` (`changeDbClusterAdvancedConfig`), write the form
  value back into `spec.proxy.config`. When `proxyConfigEnabled` is `false`,
  unset the field.

### 4. Cluster overview page

- In
  `src/pages/db-cluster-details/cluster-overview/cards/db-details/advanced-configuration/advanced-configuration.tsx`,
  add a second `<OverviewSectionRow>` for proxy configuration with the
  **engine-specific label** (same mapping as step 1), immediately
  below the existing **Parameters** (engine config) row.
- Apply the same sharding gate as the form.
- The overview's "Edit" modal
  (`edit-advanced-configuration.tsx`)
  reuses `AdvancedConfigurationForm`, so it inherits the field for free —
  just make sure `showProxyConfig` is passed in correctly.
- Add the new label to
  `cluster-overview.messages.ts`.

### 5. Wizard preview

Update
`database-preview/sections/advanced-configurations-section.tsx`
to render the proxy config preview line (same gate).

## Definition of Done

- Wizard, edit modal, overview, and preview all show the engine-specific
  proxy configuration field ("Router Configuration" for PSMDB, "PG Bouncer Configuration"
  for PG, "Proxy Configuration" for PXC) for PG and PXC, and for PSMDB **only when
  sharding is enabled**.
- Toggling the field off removes `spec.proxy.config` from the resulting CR.
- Field round-trips correctly when re-opening an existing cluster.
- New unit tests cover the schema and the read/write mapping.
- Visual review against the existing "Engine Configuration" field — the new
  field looks and behaves identically.
- E2E tests in the main test suite cover:
  - Setting proxy configuration for each engine type via the wizard
  - Verifying the field persists and displays correctly on the cluster overview
  - For PSMDB: confirm the field is hidden when sharding is disabled and shown
    when enabled

## Hints for new contributors

- The form library is **React Hook Form** (`useFormContext`). Look at the
  `engineParameters` field in the same file for a 1:1 example.
- The schema library is **Zod**. Add the two new fields next to
  `engineParameters` / `engineParametersEnabled`.
- All cross-step state lives in the wizard's RHF context, which is why the
  sharding flag is accessible from `getValues(DbWizardFormFields.sharding)`.

### PR DESCRIPTION
**Pull request by Denyme24**
### Summary

Closes #2093

- Exposes `spec.proxy.config` in the Advanced Configuration step with an engine-specific label:
  - MySQL: `Proxy Configuration`
  - PostgreSQL: `PG Bouncer Configuration`
  - MongoDB: `Router Configuration`
- Adds read/write mapping for `spec.proxy.config` in form defaults and payload builders.
- Adds UI visibility gating for MongoDB (hidden when sharding is disabled), including wizard, overview row, and edit modal.
- Preserves existing `spec.proxy.config` when updating cluster resources.

### AI Model Disclosure

Used Cursor (claude sonnet 4.6) to assist with implementation and test updates.  
Changes were self-reviewed and verified with:

```bash
pnpm test proxy-config.test.ts
pnpm exec tsc --incremental --noImplicitAny false -p .e2e/tsconfig.json
pnpm exec prettier --check .e2e/pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts
```

```
RUN  v2.1.9 /home/naman/openeverest/openeverest/ui/apps/everest

 ✓ src/components/cluster-form/advanced-configuration/proxy-config.test.ts (10 tests) 7ms

 Test Files  1 passed (1)
      Tests  10 passed (10)
   Start at  05:07:12
   Duration  7.64s
```

**Demo Video**

https://github.com/user-attachments/assets/efa31065-4998-4963-b472-c1d2ef73d327

### CODE REVIEW

**Comment by Denyme24** [comment]
the failing E2E workflow is caused by pre-existing teardown ordering in policies.e2e.ts / global teardown (dependency resources are deleted before DB clusters are fully terminated), and is unrelated to the proxy-config changes in this PR. I’m keeping this PR scoped to issue #2093 only

**Comment by Denyme24** [comment]
hey @solovevayaroslavna, would appreciate it if you could take a look when you get a chance.

File: ui/apps/everest/src/utils/db.tsx [hunk 1]
```diff
@@ -772,6 +788,7 @@ export const changeDbClusterAdvancedConfig = (
     },
     proxy: {
       ...dbCluster.spec.proxy,
+      ...(proxyConfigEnabled ? { config: proxyConfig } : { config: undefined }),
```
[the remark this answers is not in the document]
Comment (solovevayaroslavna) [reply ui/apps/everest/src/utils/db.tsx]: @recharte can't the configuration really be empty string? I thought that setting an empty string allows user to reset it to the default state

File: ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/db-details/advanced-configuration/advanced-configuration.tsx [hunk 2]
```diff
@@ -131,6 +142,12 @@ export const AdvancedConfiguration = ({
           parameters ? Messages.fields.enabled : Messages.fields.disabled
         }
       />
+      {showProxyConfig && (
+        <OverviewSectionRow
+          label={getProxyConfigLabel(dbType)}
+          content={proxyConfig || Messages.fields.disabled}
```
Comment (solovevayaroslavna) [inline ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/db-details/advanced-configuration/advanced-configuration.tsx]: If we leave everything as it is, then the visual will break in the UI, since instead of one configuration line there may be several lines. Please note how the same place is processed for Engine Parameters: 

```
<OverviewSectionRow
  label={Messages.fields.parameters}
  content={parameters ? Messages.fields.enabled : Messages.fields.disabled}
/>
```
Please adjust the display logic to enabled/disabled everywhere, as it is done in many other fields and cards to keep uniformity.

File: ui/apps/everest/src/components/cluster-form/advanced-configuration/advanced-configuration.utils.ts [hunk 3]
```diff
@@ -13,12 +14,17 @@
...
 
+import { capitalize } from '@mui/material';
 import { DbType } from '@percona/types';
 import { DbCluster, ProxyExposeType } from 'shared-types/dbCluster.types';
+import { getProxyUnitNamesFromDbType } from 'utils/db';
 import { AdvancedConfigurationFields } from './advanced-configuration.types';
 import { AdvancedConfigurationFormType } from './advanced-configuration-schema';
 import { EMPTY_LOAD_BALANCER_CONFIGURATION } from 'consts';
 
+export const getProxyConfigLabel = (dbType: DbType): string =>
```
Comment (solovevayaroslavna) [inline ui/apps/everest/src/components/cluster-form/advanced-configuration/advanced-configuration.utils.ts]: We try to follow the rule of storing all label/error/messages, etc. in separate files (`.messages.ts`). With rare exceptions, this may be broken or be part of the legacy code, so please include this function in messages file. This will help to maintain order and not increase the amount of text outside of messages.ts files.

**Review by solovevayaroslavna — CHANGES_REQUESTED** [review]

File: ui/apps/everest/src/pages/database-form/database-preview/sections/advanced-configurations-section.tsx [hunk 4]
```diff
@@ -13,9 +34,15 @@ export const AdvancedConfigurationsPreviewSection = ({
...
-}: AdvancedConfigurationType) => {
+  proxyConfigEnabled,
+  proxyConfig,
+  dbType,
+  sharding,
+}: AdvancedConfigurationsPreviewProps) => {
   const isExternalAccessEnabled =
     exposureMethod === ProxyExposeType.LoadBalancer;
+  const showProxyConfig =
+    dbType !== undefined ? !(dbType === DbType.Mongo && !sharding) : true;
```
Comment (solovevayaroslavna) [inline ui/apps/everest/src/pages/database-form/database-preview/sections/advanced-configurations-section.tsx]: I think we can do it simpler, what do you think?

```suggestion
  const showProxyConfig = dbType !== DbType.Mongo || !!sharding;
```

File: ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/db-details/advanced-configuration/advanced-configuration.tsx (hunk 2 above)
Comment (Denyme24) [reply ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/db-details/advanced-configuration/advanced-configuration.tsx]: updated the overview row to show Enabled/Disabled for proxy config

File: ui/apps/everest/src/components/cluster-form/advanced-configuration/advanced-configuration.utils.ts (hunk 3 above)
Comment (Denyme24) [reply ui/apps/everest/src/components/cluster-form/advanced-configuration/advanced-configuration.utils.ts]: moved getProxyConfigLabel into messages.ts

File: ui/apps/everest/src/pages/database-form/database-preview/sections/advanced-configurations-section.tsx (hunk 4 above)
Comment (Denyme24) [reply ui/apps/everest/src/pages/database-form/database-preview/sections/advanced-configurations-section.tsx]: yeah sure, have simplified the preview condition.

**Comment by Denyme24** [comment]
@solovevayaroslavna, thanks for reviewing, have addressed your review comments. PTAL. 🙂

**Comment by solovevayaroslavna** [comment]
@Denyme24 hey! sorry for long answer, can you please check and fix e2e tests? 

image

Proxy configuration tests should be ok, sharding test is known issue.

**Comment by Denyme24** [comment]
hey @solovevayaroslavna, have fixed the proxy config e2e tests also ran the e2e suite for proxy-config locally and all the 4 tests passes. PTAL on the logs once.
thanks

<details>
  <summary>see complete log</summary>
Running 18 tests using 1 worker

Creating test user via CLI...
✅ User 'testuser' has been created successfully

(node:215605) Warning: The 'NO_COLOR' env is ignored due to the 'FORCE_COLOR' env being set.
(Use `node --trace-warnings ...` to show where the warning was created)
  ✓  1 [session-setup] › ../setup/session-setup.ts:11:1 › Setup session user and generate sessionUser.json (1.7s)
  ✓  2 [session-setup] › ../setup/session-setup.ts:22:1 › Login Everest and generate sessionUser.json (8.5s)
  ✓  3 [session] › ../release/session/session-management.e2e.ts:47:1 › T191 - Verify token invalidity after user logout from Everest UI - Everest user (8.3s)
✅ Password for user 'testuser' has been set successfully

  ✓  4 [session] › ../release/session/session-management.e2e.ts:62:1 › T190 - Verify user is logged out and token is invalidated after updating user password (17.6s)
✅ User 'testuser' has been deleted successfully

  ✓  5 [session] › ../release/session/session-management.e2e.ts:98:1 › T189 - Verify user is logged out and token is invalidated after user deletion (19.2s)
Checking if user 'testuser' exists before deleting...
User 'testuser' does not exist. Skipping deletion.
(node:217877) Warning: The 'NO_COLOR' env is ignored due to the 'FORCE_COLOR' env being set.
(Use `node --trace-warnings ...` to show where the warning was created)
  ✓  6 [session-teardown] › ../teardown/session-teardown.ts:9:1 › Teardown session user (1.6s)
  ✓  7 [auth] › ../setup/auth.setup.ts:19:1 › Login (14.4s)
  ✓  8 [setup] › ../setup/global.setup.ts:77:1 › Backup storages (2.0s)
  ✓  9 [setup] › ../setup/global.setup.ts:157:1 › Monitoring endpoints (2.0s)
  ✓  10 [setup] › ../setup/global.setup.ts:191:1 › Close modal permanently (5.8s)

  ✓  11 [setup] › ../setup/global.setup.ts:198:1 › Save old RBAC permissions (666ms)
  ✓  12 [pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts:32:3 › Proxy Configuration field › PXC: proxy configuration field visible and persists to overview (46.0s)
  ✓  13 [pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts:69:3 › Proxy Configuration field › PostgreSQL: PG Bouncer Configuration field visible (18.1s)
  ✓  14 [pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts:85:3 › Proxy Configuration field › PSMDB: proxy config field hidden without sharding, visible with sharding (25.7s)
  ✓  15 [pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts:115:3 › Proxy Configuration field › PXC: proxy config round-trips in edit modal (52.6s)
  ✓  16 [teardown] › ../teardown/global.teardown.ts:26:3 › Teardown › Delete backup storage (4.5s)
  ✓  17 [teardown] › ../teardown/global.teardown.ts:57:3 › Teardown › Delete monitoring instances (3.9s)
  ✓  18 [teardown] › ../teardown/global.teardown.ts:86:3 › Teardown › Logout (16.8s)
::warning title=Slow Test,file=[pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts::[pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts took 2.4m
::warning title=Slow Test,file=[session] › ../release/session/session-management.e2e.ts::[session] › ../release/session/session-management.e2e.ts took 45.0s
::warning title=Slow Test,file=[teardown] › ../teardown/global.teardown.ts::[teardown] › ../teardown/global.teardown.ts took 25.2s
::notice title=🎭 Playwright Run Summary::  18 passed (4.7m)

  Slow test file: [pr] › ../pr/db-cluster/db-wizard/create-db-cluster/proxy-config.e2e.ts (2.4m)
  Slow test file: [session] › ../release/session/session-management.e2e.ts (45.0s)
  Slow test file: [teardown] › ../teardown/global.teardown.ts (25.2s)
  Consider splitting slow test files to speed up parallel execution
  18 passed (4.7m)

To open last HTML report run:

  pnpm exec playwright show-report .e2e/playwright-report

</details>

File: ui/apps/everest/src/utils/db.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Denyme24) [reply ui/apps/everest/src/utils/db.tsx]: @solovevayaroslavna, any update on this?

[the remark this answers is not in the document]
Comment (recharte) [reply ui/apps/everest/src/utils/db.tsx]: @solovevayaroslavna @Denyme24 sorry for delay in getting back to you.
You are right, in PXC (HAProxy) we set a default value if the field is an empty string. However, since this field is a plain string type in the CRD, kubernetes will interpret an absent field as an empty string. This means that not sending the field or sending an empty string is actually the same.

File: ui/apps/everest/src/utils/db.tsx [hunk 5]
```diff
@@ -786,6 +788,7 @@ export const changeDbClusterAdvancedConfig = (
...
       expose: {
```
[the remark this answers is not in the document]
Comment (solovevayaroslavna) [reply ui/apps/everest/src/utils/db.tsx:794]: @Denyme24 agree with copilot, it's not critical, as Diogo said above It's functionally the same, but the code is inconsistent.  so i would suggest to have this edit like this:
`...(proxyConfigEnabled && proxyConfig ? { config: proxyConfig } : { config: undefined }),`

File: ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/card.types.ts [hunk 6]
```diff
@@ -54,6 +55,9 @@ export type AdvancedConfigurationOverviewCardProps = {
     privateIP?: string;
     publicIP?: string;
   }[];
+  dbType: DbType;
```
[the remark this answers is not in the document]
Comment (solovevayaroslavna) [reply ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/card.types.ts]: @Denyme24 We definitely need to fix this too

**Review by solovevayaroslavna — CHANGES_REQUESTED** [review]
The review comments are addressed and e2e tests are fixed, thank you!

Just Two minor items before merging 🤗

File: ui/apps/everest/src/utils/db.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Denyme24) [reply ui/apps/everest/src/utils/db.tsx:794]: @solovevayaroslavna, yep i have made the changes. and also added a test covering the previously-untested scenario: proxyConfigEnabled=true with an empty proxyConfig string should produce config: undefined, not config: "".

File: ui/apps/everest/src/utils/db.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Denyme24) [reply ui/apps/everest/src/utils/db.tsx]: @recharte, thanks for the clarification, have made the changes.

File: ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/card.types.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Denyme24) [reply ui/apps/everest/src/pages/db-cluster-details/cluster-overview/cards/card.types.ts]: fixed. 🙂

**Comment by Denyme24** [comment]
@solovevayaroslavna, thanks for the review, i have addressed the final two minor review comments. PTAL
thanks

**Review by solovevayaroslavna — COMMENTED** [review]
Thanks again for the edits, everything looks great. I'm ready to approve, but i would like to check it last time when backend will be ready to merge 🤗

**Comment by Denyme24** [comment]
hey @solovevayaroslavna, any updates on this regarding the BE readiness?

**Review by recharte — APPROVED** [review]
Thanks @Denyme24, I just tested this for all 3 engines and it seems to be working as expected.
